### ISSUE TITLE
CMake build configuration for libsrcml that works for WebAssembly

### ISSUE THREAD

**Issue by toBeOfUse** [srcML/srcML#1966]
## TL;DR:

I have the WebAssembly pipeline generally figured out; the missing pieces are a) implementation of an option that compiles only libsrcml and not the client and b) how exactly to organize everything. You can see the code changes I've made in my fork to get WebAssembly working here: https://github.com/srcML/srcML/compare/develop...toBeOfUse:srcML:wasm-develop . I could finish this up myself and create a pull request if all of this makes sense to the maintainers.

## The Details:

Using the Emscripten compiler to produce a WebAssembly version of libsrcml requires the following:

1. The libraries used by libsrcml need to be compiled to WASM in advance. We have a working Docker setup for this.
2. We need a CMake option to build only libsrcml, not the client. The client uses libraries like libcurl which do not work well in WebAssembly (code that runs in a browser can't open arbitrary socket connections like libcurl presumably does), and WebAssembly can't be run directly from the command line, anyway. A WebAssembly version of the client could theoretically be run by manually calling its main() function from JavaScript or by using one of the newer non-browser, non-Node.js WASM runtimes, but we have decided that supporting this wouldn't be worth it. Implementing a no-client option in the develop branch would be as simple as wrapping the [`add_subdirectory(client)` command](https://github.com/srcML/srcML/blob/e5af3c1a295afd81f0ff9e3ada2ac8ad092b89fb/src/CMakeLists.txt#L195) in an if statement. (In my fork, I'm literally just skipping the client when the environment variable `EMSCRIPTEN` is true.)
3. Shared library-related commands should be avoided. When running CMake via Emscripten, even when using the option `-DBUILD_SHARED_LIBS=OFF`, CMake still complains many times that a shared library target is being created and that the current environment doesn't support those. I [fixed this](https://github.com/toBeOfUse/srcML/commit/88fbeef133de171742739396974aecd4f0a1da79) just by using `if (BUILD_SHARED_LIBS)` more in src/libsrcml/CMakeLists.txt. My version is not terribly well organized.
4. We need a new list of exported functions from the API that leaves out functions that aren't directly relevant to a WebAssembly version of the code - notably, we're leaving out the original set of convenience functions that operates on files (replacing them with a new version that operates on code strings), as well as functions that have file descriptors, FILE pointers, and C function pointers as parameters. I made [a version of the existing exports list](https://github.com/toBeOfUse/srcML/blob/a2c623c5bc485540e62b9cf0844c70384f2dac1b/src/libsrcml/export_list_emscripten) that uses comments to document with functions are needed and which aren't. This can be [easily read by CMake](https://github.com/toBeOfUse/srcML/blob/e58c19271f74f2c44cfc1cdecc46e6cd1522ed90/src/libsrcml/CMakeLists.txt#L68-L80) to pass to Emscripten via the command.
5. When compiling with Emscripten, we need to use `emcc` instead of `ld` and add a [custom POST_BUILD command](https://github.com/toBeOfUse/srcML/blob/e58c19271f74f2c44cfc1cdecc46e6cd1522ed90/src/libsrcml/CMakeLists.txt#L88-L100) that outputs WebAssembly and Javascript.

With these in place, compiling libsrcml itself to WASM is as simple as running:
```bash
emcmake cmake . -DBUILD_SHARED_LIBS=OFF -DCMAKE_BUILD_TYPE=Release  # or Debug
emmake make libsrcml_static
```

Final caveats:
- The exact command-line options that I'm using for Emscripten, which are currently hard-coded in the CMakeLists files, may end up changing slightly for optimization or other configuration purposes before all the dust has settled. It looks like there may be a better way to set those that involves CMakePreset.json.

**Comment by toBeOfUse** [comment]
And another thing: [I've added a CMake option](https://github.com/toBeOfUse/srcML/commit/bddbc380ea9229dfa8b37a02f8cdd02362aa445c) called EMSCRIPTEN_TARGET that lets one choose between compiling for "NODE" or the "BROWSER". The reason is just that when compiling for Node.js with filesystem access support, a bunch of code is emitted that the browser doesn't like. The result is cleaner code for each environment individually at the cost of some redundancy.

**Comment by toBeOfUse** [comment]
Okay, I finally did something genuinely scary. I experimentally [added four lines of code](https://github.com/toBeOfUse/srcML/commit/2e5f9fc1f5d6049914849b4899b4cd1e104e7d0c) to `srcml_archive_close`:

```C++
    if (archive->reader) {
        delete archive->reader;
        archive->reader = nullptr;
    }
```

Here's why: when an archive is opened for reading, it creates a `srcml_sax2_reader`, and stores a pointer to it in `archive->reader`. This reader immediately creates a new `std::thread`. This thread is not joined until the `srcml_sax2_reader` is destroyed, which, without my modification, only happens when the archive itself is deallocated. This basically means that one extra thread is up and running for every extant archive that has been opened for reading.

This is not a problem in native code, since threads are efficient and handled at the operating system level. However, in JavaScript in the browser, threads are implemented using Web Workers, which are thread-like but much less lightweight, and are notorious for hoarding memory. JavaScript garbage collection is not predictable enough to guarantee that archives and the threads that come with them will be cleaned up in short order (and, in fairness, garbage collectors aren't built to clean up threads.) Worse yet, the most straightforward implementation of WebAssembly threads is to create a pool of Web Workers and distribute work to them when e.g. the `std::thread` constructor is called; creating a brand new Web Worker in the middle of a call to C++ code is unperformant and requires lots of hoops, jumps, and skullduggery (potentially including other modifications to the C++, to allow the browser, which spawns Web Workers, to briefly take back control.) This is a problem, since any reasonably-sized thread pool could be completely claimed by enough open archives.

I think the cleanest solution is to join the thread by destroying the `srcml_sax2_reader` when `srcml_archive_close` is called. This seems logical, given that no more reading can be done after an archive is closed anyway. The reader will be automatically re-created if the archive is opened again. All the tests that I have for JavaScript so far pass with this method, and it allows me to easily operate with a thread pool that contains only a single worker. Importantly, all the libsrcml and client tests in srcML continue to pass as well:

![image](https://github.com/srcML/srcML/assets/49729978/d575e60a-517b-4bd4-a934-61273a547423)


![image](https://github.com/srcML/srcML/assets/49729978/2be4f1b1-1a8a-4208-8eeb-64c4755c8b7c)

**Comment by mikeweyandt** [comment]
Some recent work towards this in https://github.com/srcML/srcML/tree/issue1966 based off the current develop branch, party exploratory to understand the emscripten system

**Comment by toBeOfUse** [comment]
Make sure you have access to [this repo](https://github.com/srcML/libsrcml.js) to see the previous setup for compiling libsrcml.js to WebAssembly, and see also [this issue](https://github.com/srcML/libsrcml.js/issues/3) where I apologize for having no idea what I was doing when I created that setup
